### PR TITLE
Releases trained models for all reproduced agents

### PR DESCRIPTION
**Pull request by prabhatnagarajan**
Addresses: [#551](https://github.com/chainer/chainerrl/issues/551)

### CODE REVIEW

**Comment by muupan** [comment]
Can you fix the Travis CI errors?

File: chainerrl/misc/pretrained_models.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    It downloads a file from the URL if there is no corresponding cache. After
+    the download, this function stores a cache to the directory under the
+    dataset root (see :func:`set_dataset_root`). If there is already a cache
+    for the given URL, it just returns the path to the cache without
+    downloading the same file.
+    Args:
+        url (string): URL to download from.
+    Returns:
+        string: Path to the downloaded file.
+    """
```
Comment (muupan) [inline chainerrl/misc/pretrained_models.py]: - Why did you remove the empty lines in the docstring? We use Google Style docstrings http://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings
- It seems like you added some change to this function, e.g., 'CHAINERRL_DOWNLOAD_REPORT'. Can you change the docstring accordingly and highlight the changes from ChainerCV's function?

File: chainerrl/misc/pretrained_models.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+def download_and_store_model(alg, url, env, model_type):
+    """Downloads a model file and puts it under model directory.
+    It downloads a file from the URL and puts it under model directory.
+    If there is already a file at the destination path,
+    it just returns the path without downloading the same file.
+    Args:
+        url (string): URL to download from.
+    Returns:
+        string: Path to the downloaded file.
+    """
```
Comment (muupan) [inline chainerrl/misc/pretrained_models.py]: Ditto

File: chainerrl/misc/pretrained_models.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    "Rainbow": ["best", "final"],
+    "A3C": ["final"],
+    "DDPG": ["best", "final"],
+    "TRPO": ["best", "final"],
+    "PPO": ["final"],
+    "TD3": ["best", "final"],
+    "SAC": ["best", "final"]
+}
+
+url = "https://chainer-assets.preferred.jp/chainerrl/"
```
Comment (muupan) [inline chainerrl/misc/pretrained_models.py]: This module-level variable should be renamed as `cached_download` and `download_and_store_model` use a local variable of the same name.

File: chainerrl/misc/pretrained_models.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        if not is_cached:
+            cache_path = cached_download(os.path.join(url_basepath,
+                                                      file))
+            os.rename(cache_path, path)
+            with zipfile.ZipFile(path, 'r') as zip_ref:
+                zip_ref.extractall(root)
+        return os.path.join(root, model_type), is_cached
+
+
+def download_model(alg, env, model_type="best"):
```
Comment (muupan) [inline chainerrl/misc/pretrained_models.py]: Can you add a docstring?

**Comment by prabhatnagarajan** [comment]
To test this locally:

For each of the 9 run reproducibility scripts, run:

`python train_<alg>.py --load-pretrained --pretrained-type <final|best> --env <Env>`

Note that for some algorithms we don't have both `final` and `best` networks stored.

The link to the stored models for the tests can be received from me privately.

File: chainerrl/misc/pretrained_models.py (hunk 3 above)
Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: What do you mean? This variable should be renamed to `cached_download`? I have a function called `cached_download`. Can I call this `download_url` instead?

Comment (muupan) [reply chainerrl/misc/pretrained_models.py]: No, I meant that it should be renamed *because* `cached_download` and `download_and_store_model` use a local variable of the same name.

Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: Understood.

File: chainerrl/misc/pretrained_models.py (hunk 4 above)
Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: Addressed.

File: chainerrl/misc/pretrained_models.py (hunk 3 above)
Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: Addressed.

File: chainerrl/misc/pretrained_models.py (hunk 1 above)
Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: Addressed. I removed the `CHAINERRL_DOWNLOAD_REPORT`. I had previously walked through the ChainerCV docstring and replaced references to ChainerCV with ChainerRL. The `CHAINERRL_DOWNLOAD_REPORT ` is not relevant to CRL.

Comment (muupan) [reply chainerrl/misc/pretrained_models.py]: > The `CHAINERRL_DOWNLOAD_REPORT` is not relevant to CRL.

This function's behavior depends on `CHAINERRL_DOWNLOAD_REPORT`, so you should either document it or remove the dependency.

File: chainerrl/misc/pretrained_models.py (hunk 2 above)
Comment (muupan) [reply chainerrl/misc/pretrained_models.py]: `Args:` and `Returns:`  should be updated.

File: chainerrl/misc/pretrained_models.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    """Downloads and returns pretrained model.
+
+    Args:
+        alg (string): URL to download from.
+        env (string): Gym Environment name.
+        model_type (string): Either `best` or `final`.
+    Returns:
+        model_path: Path to the downloaded file.
```
Comment (muupan) [inline chainerrl/misc/pretrained_models.py]:
```suggestion
        str: Path to the downloaded file.
```

File: chainerrl/misc/pretrained_models.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        is_cached: whether the model was already cached.
```
Comment (muupan) [inline chainerrl/misc/pretrained_models.py]:
```suggestion
        bool: whether the model was already cached.
```

File: examples/atari/reproduction/a3c/train_a3c.py [hunk 7]
```diff
@@ -137,9 +145,9 @@ def make_env(process_idx, test):
             env=env,
             agent=agent,
             n_steps=None,
-            n_episodes=args.eval_n_runs)
-        print('n_runs: {} mean: {} median: {} stdev: {}'.format(
-            args.eval_n_runs, eval_stats['mean'], eval_stats['median'],
+            n_episodes=args.eval_n_steps)
```
Comment (muupan) [inline examples/atari/reproduction/a3c/train_a3c.py]: You meant `n_steps=args.eval_n_steps`?

File: examples/atari/reproduction/rainbow/README.md [hunk 8]
```diff
@@ -21,6 +21,21 @@ python train_rainbow.py [options]
...
+python train_rainbow.py --demo --load-pretrained --env BreakoutNoFrameskip-v4 --pretrained-type best --gpu -1
+```
+
+### Useful Options
+- `--gpu`. Specifies the GPU. If you do not have a GPU on your machine, run the example with the option `--gpu -1`.
+- `--env`. Specifies the environment. 
+- `--demo`. Runs an evaluation, instead of training the agent.
+- `--load-pretrained` Loads the pretrained model. Both `--load` and `--load-pretrained` cannot be used together.
+- `--pretrained-type`. Either `best` (the best intermediate network during training) or `final` (the final network after training).
+
```
Comment (muupan) [inline examples/atari/reproduction/rainbow/README.md]: I think `## Running the Example` and `## Running our Pretrained models` should be merged (not only for this example but also for other examples). Having two `### Useful Options`s that largely overlap is not good.

File: tests/misc_tests/test_pretrained_models.py [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+                           target_update_interval=10 ** 4,
+                           clip_delta=True,
+                           update_interval=4,
+                           batch_accumulator='sum',
+                           phi=lambda x: x)
+
+        model, exists = download_model("DQN", "BreakoutNoFrameskip-v4",
+                                       model_type=self.pretrained_type)
+        agent.load(model)
+        assert exists
```
Comment (muupan) [inline tests/misc_tests/test_pretrained_models.py]: Why existence of cache is asserted? It should be false when running this test for the first time.

File: examples/atari/reproduction/rainbow/README.md (hunk 8 above)
Comment (prabhatnagarajan) [reply examples/atari/reproduction/rainbow/README.md]: Ok, will change.

File: tests/misc_tests/test_pretrained_models.py (hunk 9 above)
Comment (prabhatnagarajan) [reply tests/misc_tests/test_pretrained_models.py]: The reason I assert for the existence of the cache is to ensure that we use the cached models during the PFN CI. When running this test locally, indeed it should be false. But when running the test on the CI, we pretrained models should already be downloaded. I omit this test from the Travis CI and only run it in the PFN CI.

Comment (muupan) [reply tests/misc_tests/test_pretrained_models.py]: Tests should be able to run locally. Can you remove the assertions?

Comment (prabhatnagarajan) [reply tests/misc_tests/test_pretrained_models.py]: Makes sense. I'll remove the assertions. But then we won't have the jobs fail if they caching of trained models isn't done properly, so we should just be vigilant and pay attention to ensure that we're not accidentally downloading models instead of using cached ones.

Comment (muupan) [reply tests/misc_tests/test_pretrained_models.py]: That should be ensured definitely, but without breaking the tests locally. I think it is good to have a env var like CHAINERRL_ASSERT_DOWNLOADED_MODEL_IS_CACHED so that we can enable the assertion optionally.

File: examples/atari/reproduction/a3c/train_a3c.py (hunk 7 above)
Comment (prabhatnagarajan) [reply examples/atari/reproduction/a3c/train_a3c.py]: Yep. Fixed.

File: tests/misc_tests/test_pretrained_models.py (hunk 9 above)
Comment (prabhatnagarajan) [reply tests/misc_tests/test_pretrained_models.py]: Done.

File: chainerrl/misc/pretrained_models.py (hunk 1 above)
Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: I removed the dependency.

File: chainerrl/misc/pretrained_models.py (hunk 2 above)
Comment (prabhatnagarajan) [reply chainerrl/misc/pretrained_models.py]: Done.
